### PR TITLE
generators/templates: Add locking to templates

### PR DESCRIPTION
**Pull request by erikh**
cc @jojimt

### CODE REVIEW

File: generators/templates.go [hunk 1]
```diff
@@ -802,22 +827,30 @@ func Create{{ initialCap .Name }}(obj *{{ initialCap .Name }}) error {
 		}
 
 		// save the original object after update
+    collections.{{ .Name }}Mutex.Lock()
 		saveObj = collections.{{ .Name }}s[obj.Key]
+    collections.{{ .Name }}Mutex.Unlock()
```
Comment (jojimt) [inline generators/templates.go:828]: Can we just use defer for the unlock (unless there is a specific reason).

File: generators/templates.go [hunk 2]
```diff
@@ -419,14 +429,18 @@ type {{ initialCap .Name }} struct {
...
+  {{ initialCap .Name }}	map[string]modeldb.Link		` + "`" + `json:"{{ .Name }},omitempty"` + "`" + `
   {{end}}
 }
 {{ end }}
 
 {{ if .Links | len }}
 type {{ initialCap .Name }}Links struct {
-  {{ range .Links }} {{ initialCap .Name }} modeldb.Link	` + "`" + `json:"{{ .Name }},omitempty"` + "`" + `
+  {{ range .Links }} 
+  {{ .Name }}Mutex sync.Mutex
```
Comment (jojimt) [inline generators/templates.go]: I think we probably don't need this lock. If we are thinking about protecting the whole object it should be at a higher level.

File: generators/templates.go [hunk 3]
```diff
@@ -457,6 +471,7 @@ type {{ initialCap .Name }}Inspect struct {
 }
   `,
 	"propstruct": `
+{{ if eq .Type "array" }}{{ .Name }}Mutex sync.Mutex{{ end }}
```
Comment (jojimt) [inline generators/templates.go]: This can also be omitted I think.

File: generators/templates/objstruct.tmpl [hunk 4]
```diff
@@ -14,14 +14,18 @@ type {{ initialCap .Name }} struct {
...
+  {{ initialCap .Name }}	map[string]modeldb.Link		`json:"{{ .Name }},omitempty"`
   {{end}}
 }
 {{ end }}
 
 {{ if .Links | len }}
 type {{ initialCap .Name }}Links struct {
-  {{ range .Links }} {{ initialCap .Name }} modeldb.Link	`json:"{{ .Name }},omitempty"`
+  {{ range .Links }} 
+  {{ .Name }}Mutex sync.Mutex
```
Comment (jojimt) [inline generators/templates/objstruct.tmpl]: I think we don't need this mutex

File: generators/templates/routeFunc.tmpl [hunk 5]
```diff
@@ -247,7 +268,9 @@ func Find{{ initialCap .Name }}(key string) *{{ initialCap .Name }} {
 
 // Delete a {{ .Name }} object
 func Delete{{ initialCap .Name }}(key string) error {
+  collections.{{ .Name }}Mutex.Lock()
```
Comment (jojimt) [inline generators/templates/routeFunc.tmpl:271]: Would be better to just use defer x.Unlock()

File: generators/templates.go (hunk 2 above)
Comment (erikh) [reply generators/templates.go]: you're right, I will prune it.

**Comment by erikh** [comment]
the defers were the original approach; the finer-grained locking was required when I started to run into some deadlocks in the tests. This is uglier but I'm pretty confident it's good now.

(just wanted to address the comments generally)

File: generators/templates/objstruct.tmpl (hunk 4 above)
Comment (erikh) [reply generators/templates/objstruct.tmpl]: eyah this is already handled at the higher levels.

**Comment by jojimt** [comment]
LGTM

**Pull request merged**